node/pkg/query/query.go

```diff
@@ -0,0 +1,412 @@
+					continue
+				}
+
+				pq, exists := pendingQueries[resp.RequestID]
+				if !exists {
+					qLogger.Warn("received a success response with no outstanding query, dropping it", zap.String("requestID", resp.RequestID), zap.Int("requestIdx", resp.RequestIdx))
+					continue
+				}
+
+				if resp.RequestIdx >= len(pq.responses) {
+					qLogger.Error("received a response with an invalid index", zap.String("requestID", resp.RequestID), zap.Int("requestIdx", resp.RequestIdx))
+					continue
+				}
+
+				// Store the result, which will mark this per-chain query as completed.
+				pq.responses[resp.RequestIdx] = resp
+
+				// If we still have other outstanding per chain queries for this request, keep waiting.
+				numStillPending := pq.numPendingRequests()
+				if numStillPending > 0 {
+					qLogger.Info("received a per chain query response, still waiting for more", zap.String("requestID", resp.RequestID), zap.Int("requestIdx", resp.RequestIdx), zap.Int("numStillPending", numStillPending))
+					continue
+				} else {
+					qLogger.Info("received final per chain query response, ready to publish", zap.String("requestID", resp.RequestID), zap.Int("requestIdx", resp.RequestIdx))
+				}
+
+				// Build the list of per chain response publications and the overall query response publication.
+				responses := []*PerChainQueryResponse{}
+				for _, resp := range pq.responses {
+					if resp == nil {
+						qLogger.Error("unexpected null response in pending query!", zap.String("requestID", resp.RequestID), zap.Int("requestIdx", resp.RequestIdx))
+						continue
+					}
+
+					responses = append(responses, &PerChainQueryResponse{
+						ChainId:  resp.ChainId,
+						Response: resp.Response,
+					})
+				}
+
+				respPub := &QueryResponsePublication{
+					Request:           pq.signedRequest,
+					PerChainResponses: responses,
+				}
+
+				// Send the response to be published.
+				select {
+				case queryResponseWriteC <- respPub:
+					qLogger.Info("forwarded query response to p2p", zap.String("requestID", resp.RequestID))
+					delete(pendingQueries, resp.RequestID)
+				default:
+					qLogger.Warn("failed to publish query response to p2p, will retry publishing next interval", zap.String("requestID", resp.RequestID))
+					pq.respPub = respPub
+				}
+			} else if resp.Status == QueryRetryNeeded {
+				if _, exists := pendingQueries[resp.RequestID]; exists {
+					qLogger.Warn("query failed, will retry next interval", zap.String("requestID", resp.RequestID), zap.Int("requestIdx", resp.RequestIdx))
+				} else {
+					qLogger.Warn("received a retry needed response with no outstanding query, dropping it", zap.String("requestID", resp.RequestID), zap.Int("requestIdx", resp.RequestIdx))
+				}
+			} else if resp.Status == QueryFatalError {
+				qLogger.Warn("received a fatal error response, dropping the whole request", zap.String("requestID", resp.RequestID), zap.Int("requestIdx", resp.RequestIdx))
```

Comment:
Does this happen a lot and that is why it is a Warn instead of an Error?

Reply:
I changed all the warnings where we give up on the request to errors.